src/System.Linq/tests/SingleTests.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System.Linq.Tests.Helpers;
+using Xunit;
+
+namespace System.Linq.Tests
+{
+    public class SingleTests
+    {
+        [Fact]
+        public void FailOnEmpty()
+        {
+            Assert.Throws<InvalidOperationException>(() => Enumerable.Empty<int>().Single());
+        }
+
+        [Fact]
+        public void DefaultOnEmpty()
+        {
+            Assert.Equal(0, Enumerable.Empty<int>().SingleOrDefault());
+        }
+
+        [Fact(Skip="Runs for thousands of years. Then throws wrong exception")]
+        public void FailOnManyMatches()
+        {
+            Assert.Throws<InvalidOperationException>(() => Enumerable.Range(0, int.MaxValue).Single(i => i > 0));
+        }
+
+        [Fact(Skip = "Runs for thousands of years. Then throws wrong exception")]
+        public void DefaultFailOnManyMatches()
+        {
+            Assert.Throws<InvalidOperationException>(() => Enumerable.Range(0, int.MaxValue).SingleOrDefault(i => i > 0));
+        }
+
+        [Fact]
+        public void FindSingle()
+        {
+            Assert.Equal(42, Enumerable.Repeat(42, 1).Single());
+        }
+
+        [Theory]
+        [InlineData(1, 100)]
+        [InlineData(42, 100)]
+        public void FindSingleMatch(int target, int range)
+        {
+            Assert.Equal(target, Enumerable.Range(0, range).Single(i => i == target));
```

Comment:
Similarly here, it be good to track the number of times the delegate is invoked and ensure it matches range.